contracts/registries/GroupIPAssetRegistry.sol

```diff
@@ -0,0 +1,182 @@
+        mapping(address ipId => address rewardPool) rewardPools;
+        // whitelisted group reward pools
+        mapping(address rewardPool => bool isWhitelisted) whitelistedGroupRewardPools;
+    }
+
+    // keccak256(abi.encode(uint256(keccak256("story-protocol.GroupIPAssetRegistry")) - 1)) & ~bytes32(uint256(0xff));
+    bytes32 private constant GroupIPAssetRegistryStorageLocation =
+        0x34c89140582ad641fa679f955c67d1a82028bef0953ade7c28b8194cf080d600;
+
+    modifier onlyGroupingModule() {
+        if (msg.sender != address(GROUPING_MODULE)) {
+            revert Errors.GroupIPAssetRegistry__CallerIsNotGroupingModule(msg.sender);
+        }
+        _;
+    }
+
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor(address groupingModule) {
+        GROUPING_MODULE = IGroupingModule(groupingModule);
+        _disableInitializers();
+    }
+
+    /// @notice Registers a Group IPA
+    /// @param groupNft The address of the group IPA
+    /// @param groupNftId The id of the group IPA
+    /// @param rewardPool The address of the group reward pool
+    /// @return groupId The address of the newly registered Group IPA.
+    function registerGroup(
+        address groupNft,
+        uint256 groupNftId,
+        address rewardPool
+    ) external nonReentrant onlyGroupingModule whenNotPaused returns (address groupId) {
+        groupId = _register({ chainid: block.chainid, tokenContract: groupNft, tokenId: groupNftId });
+
+        IIPAccount(payable(groupId)).setBool("GROUP_IPA", true);
+        GroupIPAssetRegistryStorage storage $ = _getGroupIPAssetRegistryStorage();
+        if (!$.whitelistedGroupRewardPools[rewardPool]) {
+            revert Errors.GroupIPAssetRegistry__GroupRewardPoolNotRegistered(rewardPool);
+        }
+        _getGroupIPAssetRegistryStorage().rewardPools[groupId] = rewardPool;
+    }
+
+    /// @notice Whitelists a group reward pool
+    /// @param rewardPool The address of the group reward pool
+    function whitelistGroupRewardPool(address rewardPool) external onlyGroupingModule whenNotPaused {
+        if (rewardPool == address(0)) {
+            revert Errors.GroupIPAssetRegistry__InvalidGroupRewardPool(rewardPool);
+        }
+        _getGroupIPAssetRegistryStorage().whitelistedGroupRewardPools[rewardPool] = true;
+    }
+
+    /// @notice Adds a member to a Group IPA
+    /// @param groupId The address of the Group IPA.
+    /// @param ipIds The addresses of the IPs to add to the Group IPA.
+    function addGroupMember(address groupId, address[] calldata ipIds) external onlyGroupingModule whenNotPaused {
+        if (!_isRegisteredGroup(groupId)) {
+            revert Errors.GroupIPAssetRegistry__NotRegisteredGroupIP(groupId);
+        }
+        GroupIPAssetRegistryStorage storage $ = _getGroupIPAssetRegistryStorage();
+        EnumerableSet.AddressSet storage allMemberIpIds = $.groups[groupId];
+        for (uint256 i = 0; i < ipIds.length; i++) {
+            address ipId = ipIds[i];
```

Comment:
can remove this line: `address ipId = ipIds[i];` - ipIds is external and calldata so there should not be much gas difference